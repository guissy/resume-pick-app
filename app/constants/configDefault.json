[
  {
    "name": "核心项",
    "children": [
      {
        "name": "RN",
        "alias": ["react native", "rn", "weex"],
        "score": 1.5
      },
      {
        "name": "flutter",
        "alias": ["flutter1", "flutter2", "dart"],
        "score": 1.5
      },
      {
        "name": "h5",
        "alias": [
          "webview",
          "uni-app",
          "uniapp",
          "umi-app",
          "umijs",
          "ionic",
          "apicloud",
          "mobile",
          "hybird",
          "移动端"
        ],
        "score": 1.5
      },
      {
        "name": "react",
        "alias": ["reactjs", "redux", "antd", "ant design"],
        "score": 1.5
      },
      {
        "name": "vue",
        "alias": ["vue2", "vuejs", "vue2", "element", "vant"],
        "score": 1.5
      },
      {
        "name": "vue3",
        "score": 1.5
      },
      {
        "name": "angular",
        "alias": [
          "ng",
          "ng2",
          "angularjs",
          "angular2",
          "angular4",
          "angular5",
          "angular6",
          "angular7",
          "angular8"
        ],
        "score": 1
      },
      {
        "name": "wxapp",
        "alias": ["微信", "小程序", "taro", "mpvue", "mpvue"],
        "score": 1.5
      },
      {
        "name": "node",
        "alias": ["nodejs", "express", "koa", "koa2", "egg", "next", "pm2"],
        "score": 1.5
      },
      {
        "name": "state mgr",
        "alias": [
          "redux",
          "saga",
          "mobx",
          "vuex",
          "vuetify",
          "dva",
          "事件总线",
          "eventBus",
          "event bus",
          "mvvm"
        ],
        "score": 1
      },
      {
        "name": "es6",
        "alias": [
          "es5",
          "es6",
          "es6+",
          "es2015",
          "es2016",
          "es2017",
          "es7",
          "es8",
          "es9",
          "ecmascript",
          "原生JS",
          "原生 JS"
        ],
        "score": 0.8
      },
      {
        "name": "w3c",
        "score": 0.5,
        "alias": ["mdn"]
      },
      {
        "name": "css3",
        "alias": [
          "flex",
          "grid",
          "css3",
          "rem",
          "less",
          "sass",
          "scss",
          "css-modules",
          "css in js",
          "css-in-js",
          "styled-components",
          "jss",
          "emotion",
          "tailwind",
          "纯CSS",
          "vw",
          "vh"
        ],
        "score": 0.5
      },
      {
        "name": "keyframes",
        "alias": [
          "animate.css",
          "wow.js",
          "gsap",
          "pixijs",
          "svga",
          "createjs"
        ],
        "score": 0.8
      }
    ]
  },
  {
    "name": "进阶",
    "children": [
      {
        "name": "typescript",
        "alias": ["ts"],
        "score": 1.5
      },
      {
        "name": "hooks",
        "alias": ["hook", "hoc"],
        "score": 1.5
      },
      {
        "name": "rxjs",
        "alias": ["ReactiveX", "ngrx"],
        "score": 1.5
      },
      {
        "name": "graphql",
        "alias": ["gql"],
        "score": 1.5
      },
      {
        "name": "test",
        "calc": "max",
        "alias": [
          "mocha",
          "jasmine",
          "jest",
          "cypress",
          "puppeteer",
          "unit test"
        ],
        "score": 1.5
      },
      {
        "name": "gayhub",
        "alias": ["github.com", "gitee.com"],
        "score": 1.5
      },
      {
        "name": "lint",
        "alias": [
          "tslint",
          "eslint",
          "jslint",
          "husky",
          "prettier",
          "命名规范",
          "代码规范",
          "commit规范",
          "接口规范"
        ],
        "score": 1
      },
      {
        "name": "review",
        "alias": ["pr", "git flow", "代码审查"],
        "score": 0.5
      },
      {
        "name": "blog",
        "alias": [
          "个人博客",
          "博客地址",
          "csdn.net",
          "cnblogs.com",
          "juejin.im",
          "zhihu.com",
          "jianshu.com",
          "segmentfault.com",
          "oschina.net",
          "imooc.com",
          "51cto.com",
          "gitbooks.io"
        ],
        "score": 1.5
      }
    ]
  },
  {
    "name": "其他",
    "children": [
      {
        "name": "大厂背景",
        "alias": [
          "\\b阿里巴巴\\b",
          "\\b腾讯\\b",
          "\\b百度",
          "\\b京东\\b",
          "\\b美团点评\\b",
          "\\b字节跳动\\b"
        ],
        "score": 2.5
      },
      {
        "name": "leader",
        "alias": ["组长", "负责人", "选型", "架构设计"],
        "score": 2
      },
      {
        "name": "manager",
        "alias": ["主持会议", "分配工作", "招聘"],
        "score": 2
      },
      {
        "name": "流量",
        "alias": ["([0-9一二三四五六七八九几十百千]+)\\s*(万|亿)"],
        "score": 2
      },
      {
        "name": "百分点",
        "alias": ["(减少|降低|提升|提高)(\\D{0,2})\\d+(\\.\\d+)?%"],
        "score": 2
      },
      {
        "name": "游戏引擎",
        "alias": ["白鹭", "p2", "pomelo", "cocos2d", "egret", "重力", "粒子"],
        "score": 1.5
      },
      {
        "name": "sql",
        "alias": [
          "sqlite3",
          "sqlite",
          "mysql",
          "postgre",
          "mariadb",
          "hive",
          "redis"
        ],
        "score": 1
      },
      {
        "name": "electron",
        "alias": ["nwjs"],
        "score": 1.2
      },
      {
        "name": "mock",
        "alias": ["mockjs", "rap", "模拟", "抓取"],
        "score": 0.5
      },
      {
        "name": "ejs",
        "alias": ["hygen", "swagger"],
        "score": 0.5
      },
      {
        "name": "git",
        "score": 0.5
      },
      {
        "name": "docker",
        "score": 0.5
      },
      {
        "name": "websocket",
        "alias": ["socket", "canvas", "charles"],
        "score": 0.5
      },
      {
        "name": "微服务",
        "alias": ["monorepo", "submodule", "无服务器", "severless"],
        "score": 0.5
      },
      {
        "name": "加密",
        "alias": [
          "密文",
          "支付",
          "md5",
          "sha1",
          "签名",
          "白名单",
          "jwt",
          "x-token",
          "crypto"
        ],
        "score": 0.5
      },
      {
        "name": "重构",
        "alias": [
          "防抖",
          "去抖",
          "节流",
          "虚拟化",
          "懒加载",
          "debounce",
          "throttle",
          "代码分割",
          "分包",
          "错误收集",
          "埋点",
          "preact"
        ],
        "score": 0.5
      },
      {
        "name": "Promise",
        "alias": ["async", "await", "generator"],
        "score": 0.5
      },
      {
        "name": "自定义",
        "alias": [
          "封装",
          "公共方法",
          "二次封装",
          "造轮子",
          "简化",
          "可复用",
          "自动化"
        ],
        "score": 0.5
      },
      {
        "name": "性能调优",
        "alias": ["毫秒", "攻克"],
        "score": 0.5
      },
      {
        "name": "component",
        "alias": ["基础组件", "公共组件", "通用组件", "业务组件", "组件化"],
        "score": 0.5
      },
      {
        "name": "编程范式",
        "alias": ["函数式", "haskell", "f#"],
        "score": 0.5
      },
      {
        "name": "技术博客",
        "alias": [
          "极客",
          "掘金",
          "知乎",
          "思否",
          "stackoverflow",
          "mdn",
          "w3school",
          "博客园",
          "简书",
          "CSDN"
        ],
        "score": 0.5
      },
      {
        "name": "python",
        "score": 0.5
      },
      {
        "name": "java",
        "alias": ["springcloud", "springboot", "springmvc", "mybatis"],
        "score": 0.5
      },
      {
        "name": "middleware",
        "alias": ["nacos", "eureka", "hystrix", "gateway", "zuul", "feign"],
        "score": 0.5
      },
      {
        "name": "queue",
        "alias": ["rocketmq", "rabbitmq"],
        "score": 0.5
      },
      {
        "name": "go",
        "score": 0.5
      },
      {
        "name": "rust",
        "score": 0.5
      },
      {
        "name": "cpp",
        "score": 0.5
      },
      {
        "name": "swift",
        "score": 0.5
      },
      {
        "name": "kotlin",
        "score": 0.5
      },
      {
        "name": "scala",
        "score": 0.5
      },
      {
        "name": "mongodb",
        "alias": ["sequelize", "typeorm", "orm"],
        "score": 0.5
      },
      {
        "name": "devops",
        "alias": ["shell", "nginx", "linux", "cicd"],
        "score": 0.5
      },
      {
        "name": "vim",
        "score": 0.5
      },
      {
        "name": "markdown",
        "score": 0.5
      },
      {
        "name": "ssr",
        "score": 0.5
      },
      {
        "name": "bi",
        "score": 0.5,
        "alias": ["可视化", "大屏", "echarts", "d3", "antv"]
      },
      {
        "name": "three.js",
        "score": 0.5
      },
      {
        "name": "postman",
        "score": 0.5,
        "alias": ["charles", "fiddler"]
      },
      {
        "name": "leetcode",
        "score": 0.5,
        "alias": ["力扣", "leet code"]
      },
      {
        "name": "deno",
        "score": 0.5
      },
      {
        "name": "wasm",
        "score": 0.5,
        "alias": ["web assembly"]
      },
      {
        "name": "代表作",
        "score": 0.5,
        "alias": ["项目链接", "作品链接", "(由|从)(0|零)到(1|一)"]
      },
      {
        "name": "iframe",
        "alias": ["flash", "dreamweaver", "div+css", "v-for", "v-if"],
        "score": -1.5
      },
      {
        "name": "mvc",
        "score": -0.5
      },
      {
        "name": "ajax",
        "alias": ["ajax", "jsonp", "ie6", "ie7", "ie8"],
        "score": -1
      },
      {
        "name": "jquery",
        "alias": [
          "jq",
          "zepto",
          "knockout",
          "knockoutjs",
          "extjs",
          "backbone",
          "dojo",
          "mootools",
          "emberjs",
          "yui3",
          "prototype.js"
        ],
        "score": -0.5
      },
      {
        "name": "svn",
        "alias": [
          "requirejs",
          "gulp",
          "grunt",
          "yeo",
          "yeoman",
          "seajs",
          "phonegap",
          "jasmine"
        ],
        "score": -1
      },
      {
        "name": "积极乐观",
        "alias": ["为人和蔼", "吃苦耐劳", "上进", "了解"],
        "score": -0.5
      },
      {
        "name": "熟悉",
        "score": -0.3
      }
    ]
  }
]
