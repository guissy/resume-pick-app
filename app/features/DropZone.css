.mainBox {
  flex: 1;
  min-height: 90px;
  margin-top: 20px;
  border: 1px dashed #ddd;
  cursor: pointer;
}
.dropBox {
  width: 100%;
  height: 100%;
  padding: 30px;
  display: flex;
  justify-content: center;
  align-content: center;
  box-sizing: border-box;
}

.placeholder {
  width: 100%;
  color: #eee;
}

.textCenter {
  text-align: center;
}

.upload {
  color: #666;
}

.mainBox:hover .upload {
  color: #aaa;
}
