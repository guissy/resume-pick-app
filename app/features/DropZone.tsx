import React from 'react';
// Import the useDropzone hooks from react-dropzone
import { DropzoneProps, useDropzone } from 'react-dropzone';
import styles from './DropZone.css';

const DropZone = ({ onDrop, accept }: Partial<DropzoneProps>) => {
  // Initializing useDropzone hooks with options
  const { getRootProps, getInputProps, isDragActive } = useDropzone({
    onDrop,
    accept,
  });

  /*
    useDropzone hooks exposes two functions called getRootProps and getInputProps
    and also exposes isDragActive boolean
  */

  return (
    // eslint-disable-next-line react/jsx-props-no-spreading
    <div {...getRootProps()} className={styles.mainBox}>
      {/* eslint-disable-next-line react/jsx-props-no-spreading */}
      <input className="dropzone-input" {...getInputProps()} />
      <div className={styles.dropBox}>
        <section className={styles.placeholder}>
          {isDragActive ? (
            <p className={styles.textCenter}>文件放在这里</p>
          ) : (
            <p className={styles.textCenter}>
              {/* eslint-disable-next-line react/no-unescaped-entities */}
              拖拽文件到这里
            </p>
          )}
        </section>
      </div>
    </div>
  );
};

export default DropZone;
