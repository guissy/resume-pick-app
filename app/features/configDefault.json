[
  {
    "name": "核心项",
    "children": [
      {
        "name": "RN",
        "alias": ["react native", "rn", "weex"],
        "score": 1.5
      },
      {
        "name": "flutter",
        "alias": ["flutter", "dart"],
        "score": 1.5
      },
      {
        "name": "h5",
        "alias": [
          "webview",
          "uni-app",
          "uniapp",
          "umi-app",
          "ionic",
          "apicloud",
          "mobile",
          "hybird"
        ],
        "score": 1.5
      },
      {
        "name": "react",
        "alias": ["redux", "antd", "ant design"],
        "score": 1.5
      },
      {
        "name": "vue",
        "alias": ["vue2", "vuejs", "vue2", "element", "vant"],
        "score": 1.5
      },
      {
        "name": "vue3",
        "score": 1.5
      },
      {
        "name": "angular",
        "alias": [
          "ng",
          "ng2",
          "angular2",
          "angular4",
          "angular5",
          "angular6",
          "angular7",
          "angular8"
        ],
        "score": 1
      },
      {
        "name": "wxapp",
        "alias": ["微信", "小程序", "taro", "mpvue", "mpvue"],
        "score": 1.5
      },
      {
        "name": "node",
        "alias": ["nodejs", "express", "keo", "egg", "next", "typeorm", "pm2"],
        "score": 1.5
      },
      {
        "name": "electron",
        "alias": ["nwjs"],
        "score": 1.5
      },
      {
        "name": "state mgr",
        "alias": [
          "redux",
          "saga",
          "mobx",
          "vuex",
          "vuetify",
          "dva",
          "事件总线",
          "eventBus",
          "event bus"
        ],
        "score": 1
      },
      {
        "name": "es6",
        "alias": [
          "es5",
          "es6+",
          "es2015",
          "es2016",
          "es2017",
          "es7",
          "es8",
          "es9",
          "ecmascript"
        ],
        "score": 0.5
      },
      {
        "name": "w3c",
        "score": 0.5,
        "alias": ["mdn"]
      },
      {
        "name": "css3",
        "alias": [
          "flex",
          "grid",
          "css3",
          "rem",
          "less",
          "sass",
          "css-modules",
          "css in js",
          "css-in-js",
          "styled-components",
          "jss",
          "emotion"
        ],
        "score": 0.5
      }
    ]
  },
  {
    "name": "进阶",
    "children": [
      {
        "name": "typescript",
        "alias": ["ts"],
        "score": 1.5
      },
      {
        "name": "rxjs",
        "alias": ["ReactiveX", "ngrx"],
        "score": 1.5
      },
      {
        "name": "graphql",
        "alias": ["gql"],
        "score": 1.5
      },
      {
        "name": "test",
        "calc": "max",
        "alias": [
          "mocha",
          "jasmine",
          "jest",
          "cypress",
          "puppeteer",
          "unit test"
        ],
        "score": 1.5
      },
      {
        "name": "lint",
        "alias": ["tslint", "eslint", "jslint", "命名规范", "代码规范"],
        "score": 0.5
      },
      {
        "name": "review",
        "alias": ["pr", "git flow", "代码审查"],
        "score": 0.5
      },
      {
        "name": "gayhub",
        "alias": ["github.com", "gitee.com"],
        "score": 1.5
      },
      {
        "name": "blog",
        "alias": [
          "个人博客",
          "博客地址",
          "csdn.net",
          "cnblogs.com",
          "juejin.im",
          "zhihu.com",
          "jianshu.com",
          "segmentfault.com",
          "oschina.net",
          "imooc.com",
          "51cto.com",
          "gitbooks.io"
        ],
        "score": 1.5
      }
    ]
  },
  {
    "name": "其他",
    "children": [
      {
        "name": "leader",
        "alias": ["组长", "负责人"],
        "score": 1
      },
      {
        "name": "mock",
        "score": 0.5
      },
      {
        "name": "git",
        "score": 0.5
      },
      {
        "name": "docker",
        "score": 0.5
      },
      {
        "name": "websocket",
        "alias": ["socket", "canvas", "charles"],
        "score": 0.5
      },
      {
        "name": "微服务",
        "score": 0.5
      },
      {
        "name": "加密",
        "alias": ["密文", "支付", "md5", "sha1", "签名", "白名单"],
        "score": 0.5
      },
      {
        "name": "重构",
        "alias": ["防抖", "去抖", "节流", "虚拟化", "懒加载"],
        "score": 0.5
      },
      {
        "name": "自定义",
        "alias": ["封装", "公共组件", "二次封装"],
        "score": 0.5
      },
      {
        "name": "python",
        "score": 0.5
      },
      {
        "name": "java",
        "score": 0.5
      },
      {
        "name": "go",
        "score": 0.5
      },
      {
        "name": "rust",
        "score": 0.5
      },
      {
        "name": "cpp",
        "score": 0.5
      },
      {
        "name": "swift",
        "score": 0.5
      },
      {
        "name": "kotlin",
        "score": 0.5
      },
      {
        "name": "scala",
        "score": 0.5
      },
      {
        "name": "mysql",
        "score": 0.5
      },
      {
        "name": "mongodb",
        "score": 0.5
      },
      {
        "name": "electron",
        "score": 0.5
      },
      {
        "name": "devops",
        "alias": ["shell", "nginx", "linux", "cicd"],
        "score": 0.5
      },
      {
        "name": "vim",
        "score": 0.5
      },
      {
        "name": "ssr",
        "score": 0.5
      },
      {
        "name": "bi",
        "score": 0.5,
        "alias": ["可视化", "大屏", "echarts", "d3"]
      },
      {
        "name": "three.js",
        "score": 0.5
      },
      {
        "name": "postman",
        "score": 0.5,
        "alias": ["charles", "fiddler"]
      },
      {
        "name": "leetcode",
        "score": 0.5,
        "alias": ["力扣", "leet code"]
      },
      {
        "name": "deno",
        "score": 0.5
      },
      {
        "name": "wasm",
        "score": 0.5,
        "alias": ["web assembly"]
      },
      {
        "name": "iframe",
        "alias": ["flash", "dreamweaver", "div+css", "v-for", "v-if"],
        "score": -1.5
      },
      {
        "name": "ajax",
        "alias": ["ajax", "jsonp", "ie6", "ie7", "ie8"],
        "score": -1
      },
      {
        "name": "jquery",
        "alias": ["jq", "zepto"],
        "score": -0.5
      },
      {
        "name": "svn",
        "alias": ["requirejs", "gulp", "grunt", "svn"],
        "score": -1
      },
      {
        "name": "积极乐观",
        "alias": ["为人和蔼", "吃苦耐劳", "上进", "了解"],
        "score": -0.5
      },
      {
        "name": "熟悉",
        "score": -0.3
      }
    ]
  }
]
