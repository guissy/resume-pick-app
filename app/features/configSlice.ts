import { createSlice } from '@reduxjs/toolkit';
// eslint-disable-next-line import/no-cycle
import { AppThunk, RootState } from '../store';
import { ConfigFile } from './type';
import initFile, { initDefaultFile, saveFile } from './configUtil';

export const configSlice = createSlice({
  name: 'config',
  initialState: { configs: {} as ConfigFile },
  reducers: {
    buildConfig: (state, action) => {
      if (Object.keys(state.configs).length === 0) {
        state.configs = action.payload;
      }
    },
    updateConfig: (state, action) => {
      Object.assign(state.configs, action.payload);
    },
    saveConfig: (state) => {
      saveFile(state.configs);
    },
    resetConfig: (state, action) => {
      state.configs = action.payload;
      saveFile(state.configs);
    },
  },
});

export const {
  updateConfig,
  buildConfig,
  saveConfig,
  resetConfig,
} = configSlice.actions;

export const initConfigAsync = (): AppThunk => (dispatch) =>
  initFile()
    .then((data) => dispatch(buildConfig(data)))
    // eslint-disable-next-line no-console
    .catch((e) => console.error(e));

export const resetConfigAsync = (): AppThunk => (dispatch) =>
  initDefaultFile()
    .then((data) => dispatch(resetConfig(data)))
    // eslint-disable-next-line no-console
    .catch((e) => console.error(e));

export default configSlice.reducer;

export const selectConfig = (state: RootState) => state.config.configs.default;
