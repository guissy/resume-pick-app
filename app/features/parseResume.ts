// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import * as timeContent from 'resume-pick/lib/time-content';
import fs from 'fs';
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import pdf from 'pdf-parse/lib/pdf-parse';
import textract from 'textract';
import { Config, Keyword } from './type';

function parseResumeText(
  path: string,
  config: Config,
  callback: (path: string, score: number, keywords: Keyword[]) => void,
  data: string
) {
  if (data && data.length > 0) {
    const { score, keywords: kw } = timeContent.calcTotal(data, config);
    callback(path, score, kw.items);
  } else {
    callback(path, 0, []);
  }
}

type PdfItem = { page: number; text: string };

export default function parseResume(
  this: unknown,
  path: string,
  config: Config,
  callback: (path: string, score: number, keywords: Keyword[]) => void
) {
  if (path.endsWith('.pdf')) {
    fs.readFile(path, (_e, buffer) => {
      pdf(buffer)
        .then((data: PdfItem) => {
          return parseResumeText(path, config, callback, data.text);
        })
        .catch(() => {
          parseResumeText(path, config, callback, '');
        });
    });
  } else {
    textract.fromFileWithPath(path, (_err: unknown, docText: string) => {
      parseResumeText(path, config, callback, docText);
    });
  }
}
