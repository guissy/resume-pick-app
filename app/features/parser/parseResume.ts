// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import * as timeContent from 'resume-pick/lib/time-content';
import fs from 'fs';
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import pdf from 'pdf-parse/lib/pdf-parse';
import textract from 'textract';
import {
  Config,
  Keyword,
  KeywordCalcResult,
  KeywordUtil,
  ParseResumeFn,
} from '../type';
import {
  calcLevelSalaryRate,
  calcSentiment,
  trackDegree,
  trackLinks,
  trackPhone,
  trackSalary,
  trackSchool,
} from './tractWorkAge';

const cache = new Map<
  string,
  {
    text: string;
    workAge: number;
    phone: string;
    keywords: KeywordUtil;
    links: string[];
    sentiment: number;
  }
>();

export function parseResumeText(
  path: string,
  config: Config,
  search: string,
  callback: ParseResumeFn,
  text: string
) {
  const phone = trackPhone(text);
  const links = trackLinks(text);
  const school = trackSchool(text);
  const degree = trackDegree(text);
  const salary = trackSalary(text);
  const configPure = config.filter((v) => v.name !== 'search');
  const searchs = search
    .trim()
    .split(' ')
    .map((n) => ({
      name: n,
      score: 2,
      alias: [],
      works: [],
      gained: 0,
      children: [],
    }));
  const configOk = search?.trim()
    ? [
        ...configPure,
        {
          name: 'search',
          children: searchs,
          score: 2,
          gained: 0,
          alias: [],
          works: [],
        } as Keyword,
      ]
    : configPure;
  const sentiment = calcSentiment(text, configOk);
  if (text && text.length > 0) {
    const {
      score,
      workAge,
      level,
      levelValue,
      keywords: kw,
    } = timeContent.calcTotal(text, configOk) as KeywordCalcResult;
    const levelSalary = calcLevelSalaryRate(salary, levelValue);
    const kws = kw.items.map((k: Keyword) => ({
      ...k,
      children: k.children
        .filter((w) => w.gained !== 0)
        .map((w) => ({
          ...w,
          works: w.works.map((x) => ({
            ...x,
            startDate: String(x.startDate),
            endDate: String(x.endDate),
          })),
        })),
    }));
    callback({
      path,
      score,
      keywords: kws,
      text,
      workAge,
      level,
      levelValue,
      levelSalary,
      school,
      degree,
      salary,
      links,
      phone,
      sentiment,
      search,
    });
  } else {
    callback({
      path,
      score: 0,
      keywords: [],
      text,
      workAge: 0,
      level: '-',
      levelValue: 0,
      levelSalary: 0,
      school,
      degree: '',
      salary: '',
      links,
      phone,
      sentiment,
      search,
    });
  }
}

type PdfItem = { page: number; text: string };

export default function parseResume(
  this: unknown,
  path: string,
  config: Config,
  search: string,
  callback: ParseResumeFn
) {
  const kwUtil = cache.get(path);
  if (!kwUtil) {
    if (path.endsWith('.pdf')) {
      fs.readFile(path, (_e, buffer) => {
        pdf(buffer)
          .then((data: PdfItem) => {
            return parseResumeText(path, config, search, callback, data.text);
          })
          .catch(() => {
            parseResumeText(path, config, search, callback, '');
          });
      });
    } else {
      textract.fromFileWithPath(
        path,
        { preserveLineBreaks: true },
        (_err: unknown, docText: string) => {
          parseResumeText(path, config, search, callback, docText);
        }
      );
    }
  } else {
    const scoreMap = new Map<string, number>();
    config.forEach((k) => k.children.map((w) => scoreMap.set(w.name, w.score)));
    // kwUtil.keywords.walk(kwUtil.keywords.items);
    kwUtil.keywords.walked = kwUtil.keywords.walked.map((w) => ({
      ...w,
      score: scoreMap.get(w.name) || 0,
      gained: (w.gained * (scoreMap.get(w.name) || 0)) / w.score,
    }));
    const score = kwUtil.keywords.calc(kwUtil.keywords.items);
    const kws = kwUtil.keywords.items.map((k: Keyword) => ({
      ...k,
      children: k.children.filter((w) => w.gained !== 0),
    }));
    callback({
      ...kwUtil,
      path,
      score,
      search,
      level: '',
      levelValue: 0,
      levelSalary: 0,
      school: '',
      degree: '',
      salary: '',
      keywords: kws,
    });
  }
}
