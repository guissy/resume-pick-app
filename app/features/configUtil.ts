// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import storage from 'electron-storage';
import configDefault from './configDefault.json';
import { Config } from './type';

function initFile(): Promise<Record<string, Config>> {
  return storage.get('keywords.json').catch((err: unknown) => {
    // eslint-disable-next-line no-console
    console.error(err);
    return storage.set('keywords.json', { default: configDefault });
  });
}

export function initDefaultFile(): Promise<Record<string, Config>> {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  return Promise.resolve({ default: configDefault });
}

export function saveFile(configs: Record<string, Config>) {
  return storage
    .set('keywords.json', JSON.stringify(configs, null, '  '))
    .catch((err: unknown) => {
      // eslint-disable-next-line no-console
      console.error(err);
    });
}

export default initFile;
