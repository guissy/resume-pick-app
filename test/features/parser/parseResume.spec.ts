import fs from 'fs';
import parseResume, {
  parseResumeText,
} from '../../../app/features/parser/parseResume';
import configDefault from '../../../app/constants/configDefault.json';
import { Config } from '../../../app/features/type';

const resumeText = '15K-20K 2019-08 ~ 2020-08 react';
describe('parseResume', () => {
  it('parseResumeText', () => {
    parseResumeText(
      'file.txt',
      (configDefault as unknown) as Config,
      '',
      ({
        score,
        workAge,
        sentiment,
        keywords,
        level,
        levelValue,
        levelSalary,
      }) => {
        expect(score).toBeGreaterThan(0);
        expect(workAge).toBeGreaterThan(0);
        expect(sentiment).toBeGreaterThan(0);
        expect(levelSalary).toBeGreaterThan(0);
        expect(levelValue).toBeGreaterThan(0);
        expect(/p\d+(\+){0,2}/.test(level)).toBeTruthy();
        expect(keywords).toMatchSnapshot();
      },
      resumeText
    );
  });
  it('parseResume', () => {
    const file = 'file.txt';
    fs.writeFileSync(file, resumeText);
    parseResume(
      file,
      (configDefault as unknown) as Config,
      '',
      ({
        score,
        workAge,
        sentiment,
        keywords,
        level,
        levelValue,
        levelSalary,
      }) => {
        expect(score).toBeGreaterThan(0);
        expect(workAge).toBeGreaterThan(0);
        expect(sentiment).toBeGreaterThan(0);
        expect(levelSalary).toBeGreaterThan(0);
        expect(levelValue).toBeGreaterThan(0);
        expect(/p\d+(\+){0,2}/.test(level)).toBeTruthy();
        expect(keywords).toMatchSnapshot();
      }
    );
  });
});
