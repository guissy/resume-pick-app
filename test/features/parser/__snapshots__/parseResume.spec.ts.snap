// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseResume parseResumeText 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "alias": Array [
          "reactjs",
          "redux",
          "antd",
          "ant design",
        ],
        "gained": 1.3443609377704335,
        "months": 12,
        "name": "react",
        "score": 1.5,
        "works": Array [
          Object {
            "endDate": "Sun Aug 30 2020 08:00:00 GMT+0800 (China Standard Time)",
            "startDate": "Thu Aug 01 2019 08:00:00 GMT+0800 (China Standard Time)",
            "workContent": " react",
          },
        ],
      },
    ],
    "gained": 1.3443609377704335,
    "name": "核心项",
  },
  Object {
    "children": Array [],
    "gained": 0,
    "name": "进阶",
  },
  Object {
    "children": Array [],
    "gained": 0,
    "name": "其他",
  },
]
`;
